### ISSUE TITLE
Part of face even no real face can be recognize as humanbeings.

### ISSUE THREAD

**Issue by Howjune** [vladmandic/human#84]
**Issue Description**
Part of face even no real face can be recognize as humanbeings.
**Steps to Reproduce**
1.Open  face detector only.
2.Fist to the camera.
3.Function of human.detect returns value which has face and the face.length>0.

**Expected Behavior**
The fucntion reutrns value no face if there is no face in the detection area.

**Environment
- Browser or NodeJS and version (e.g. NodeJS 14.15 or Chrome 86)
- OS and Hardware platform (e.g. Windows 10, Ubuntu Linux on x64, Android 10)
- Packager (if any) (e.g, webpack, rollup, parcel, esbuild, etc.)

**Additional**
```
    "@vladmandic/human": "0.9.26",
```
```
{
    backend: 'webgl', // select tfjs backend to use
    wasmPath: `/human/assets/`, // path for wasm binaries
    // only used for backend: wasm
    console: debug, // enable debugging output to console
    async: true, // execute enabled models in parallel
    // this disables per-model performance data but
    // slightly increases performance
    // cannot be used if profiling is enabled
    profile: false, // enable tfjs profiling
    // this has significant performance impact
    // only enable for debugging purposes
    // currently only implemented for age,gender,emotion models
    deallocate: false, // aggresively deallocate gpu memory after each usage
    // only valid for webgl backend and only during first call
    // cannot be changed unless library is reloaded
    // this has significant performance impact
    // only enable on low-memory devices
    scoped: false, // enable scoped runs
    // some models *may* have memory leaks,
    // this wrapps everything in a local scope at a cost of performance
    // typically not needed
    videoOptimized: true, // perform additional optimizations when input is video,
    // must be disabled for images
    // basically this skips object box boundary detection for every n frames
    // while maintaining in-box detection since objects cannot move that fast
    warmup: 'face', // what to use for human.warmup(), can be 'none', 'face', 'full'
    // warmup pre-initializes all models for faster inference but can take
    // significant time on startup
    filter: {
      enabled: false, // enable image pre-processing filters
      width: 0, // resize input width
      height: 0, // resize input height
      // if both width and height are set to 0, there is no resizing
      // if just one is set, second one is scaled automatically
      // if both are set, values are used as-is
      return: true, // return processed canvas imagedata in result
      brightness: 0, // range: -1 (darken) to 1 (lighten)
      contrast: 0, // range: -1 (reduce contrast) to 1 (increase contrast)
      sharpness: 0, // range: 0 (no sharpening) to 1 (maximum sharpening)
      blur: 0, // range: 0 (no blur) to N (blur radius in pixels)
      saturation: 0, // range: -1 (reduce saturation) to 1 (increase saturation)
      hue: 0, // range: 0 (no change) to 360 (hue rotation in degrees)
      negative: false, // image negative
      sepia: false, // image sepia colors
      vintage: false, // image vintage colors
      kodachrome: false, // image kodachrome colors
      technicolor: false, // image technicolor colors
      polaroid: false, // image polaroid camera effect
      pixelate: 0 // range: 0 (no pixelate) to N (number of pixels to pixelate)
    },

    gesture: {
      enabled: true // enable simple gesture recognition
    },

    face: {
      enabled: true, // controls if specified modul is enabled
      // face.enabled is required for all face models:
      // detector, mesh, iris, age, gender, emotion
      // (note: module is not loaded until it is required)
      detector: {
        modelPath: `/human/models/blazeface-back.json`, // can be 'front' or 'back'.
        // 'front' is optimized for large faces
        // such as front-facing camera and
        // 'back' is optimized for distanct faces.
        inputSize: 256, // fixed value: 128 for front and 256 for 'back'
        rotation: false, // use best-guess rotated face image or just box with rotation as-is
        maxFaces: 5, // maximum number of faces detected in the input
        // should be set to the minimum number for performance
        skipFrames: 11, // how many frames to go without re-running the face bounding box detector
        // only used for video inputs
        // e.g., if model is running st 25 FPS, we can re-use existing bounding
        // box for updated face analysis as the head probably hasn't moved much
        // in short time (10 * 1/25 = 0.25 sec)
        minConfidence: 0.5, // threshold for discarding a prediction
        iouThreshold: 0.2, // threshold for deciding whether boxes overlap too much in
        // non-maximum suppression (0.1 means drop if overlap 10%)
        scoreThreshold: 0.5 // threshold for deciding when to remove boxes based on score
        // in non-maximum suppression,
        // this is applied on detection objects only and before minConfidence
      },

      mesh: {
        enabled: false,
        modelPath: `/human/models/facemesh.json`,
        inputSize: 192, // fixed value,
        returnRawData: false // in addition to standard mesh and box values, return raw normalized values as well
      },

      iris: {
        enabled: false,
        modelPath: `/human/models/iris.json`,
        inputSize: 64 // fixed value
      },

      age: {
        enabled: false,
        modelPath: `${resourcesUrl}human/models/age-ssrnet-imdb.json`, // can be 'age-ssrnet-imdb' or 'age-ssrnet-wik'
        // which determines training set for model
        inputSize: 64, // fixed value
        skipFrames: 31 // how many frames to go without re-running the detector
        // only used for video inputs
      },

      gender: {
        enabled: false,
        minConfidence: 0.1, // threshold for discarding a prediction
        modelPath: `/human/models/gender-ssrnet-imdb.json`, // can be 'gender', 'gender-ssrnet-imdb' or 'gender-ssrnet-wik'
        inputSize: 64, // fixed value
        skipFrames: 41 // how many frames to go without re-running the detector
        // only used for video inputs
      },

      emotion: {
        enabled: false,
        inputSize: 64, // fixed value
        minConfidence: 0.2, // threshold for discarding a prediction
        skipFrames: 21, // how many frames to go without re-running the detector
        modelPath: `/human/models/emotion-large.json` // can be 'mini', 'larg'
      },

      embedding: {
        enabled: false,
        inputSize: 112, // fixed value
        modelPath: `/human/models/mobilefacenet.json`
      }
    },

    body: {
      enabled: false,
      modelPath: `/human/models/posenet.json`,
      inputSize: 257, // fixed value
      maxDetections: 10, // maximum number of people detected in the input
      // should be set to the minimum number for performance
      scoreThreshold: 0.5, // threshold for deciding when to remove boxes based on score
      // in non-maximum suppression
      nmsRadius: 20, // radius for deciding points are too close in non-maximum suppression
      outputStride: 16, // size of block in which to run point detectopn, smaller value means higher resolution
      // defined by model itself, can be 8, 16, or 32
      modelType: 'MobileNet' // Human includes MobileNet version, but you can switch to ResNet
    },

    hand: {
      enabled: false,
      inputSize: 256, // fixed value
      skipFrames: 15, // how many frames to go without re-running the hand bounding box detector
      // only used for video inputs
      // e.g., if model is running st 25 FPS, we can re-use existing bounding
      // box for updated hand skeleton analysis as the hand probably
      // hasn't moved much in short time (10 * 1/25 = 0.25 sec)
      minConfidence: 0.5, // threshold for discarding a prediction
      iouThreshold: 0.1, // threshold for deciding whether boxes overlap too much
      // in non-maximum suppression
      scoreThreshold: 0.8, // threshold for deciding when to remove boxes based on
      // score in non-maximum suppression
      maxHands: 1, // maximum number of hands detected in the input
      // should be set to the minimum number for performance
      landmarks: true, // detect hand landmarks or just hand boundary box
      detector: {
        modelPath: `/human/models/handdetect.json`
      },
      skeleton: {
        modelPath: `/human/models/handskeleton.json`
      }
    }
  };

```

Looking forward to ur reply.
Tks.

**Comment by vladmandic** [comment]
how it works is that detector prepares list of candidates for face boxes and then mesh actually discards invalid entries and also stabilizes facial bounding boxes  
if mesh is disabled, output of detector only is not reliable  

can you retry with mesh enabled,  
but instead of just looking at `face.length > 0`, check the values for 
```js
for (const f in face) console.log(`Face #{f} ${face[f].faceConfidence} ${face[f].boxConfidence}`);
```

i'll update docs to reflect this

**Comment by Howjune** [comment]
Thank you for reply.
The resaon I set the mesh disabled is in order to save some flow.
I 'll try it now.

**Comment by Howjune** [comment]
So,I found my fist scored 0.8330078125.Sometimes my face scores below that.

**Comment by Howjune** [comment]
And I found the type of human in node_modules\@vladmandic\human\dist change to typescript.

**Comment by vladmandic** [comment]
> So,I found my fist scored 0.8330078125 and sometimes my face scores below that.

That's the `boxConfidence` value? Because `faceConfidence` (when mesh is enabled) should be far lower for the fist :)

> And I found the type of human in node_modules@vladmandic\human\dist change to typescript.

Yes, I've switched the project source to TS. But the stuff in /dist is always compiled JS.  

Let me know if you have any other questions...

**Comment by Howjune** [comment]
That's just result.face.confidence then disable the mesh. And I also got the faceConfidence and boxConfidence which is high score when I use my fist to the camera.

**Comment by vladmandic** [comment]
`result.face.confidence` is just a pointer to `boxConfidence` if mesh is disabled or to `faceConfidence` if mesh is enabled.

**Comment by Howjune** [comment]
I found a way around this problem by double-checking the returned data, but it doesn't work on Firefox and will report an error.
the error is :
```
20:15:35.556 Human: Human: invalid input 
<img src="data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD…">
log.js:6:19
20:15:35.563 Human: could not convert input to tensor
```

**Comment by Howjune** [comment]
```
          const imgRepeated = document.createElement('img');
          if (photo) {
            imgRepeated.src = photo;
            const fianlResult = await this.human.detect(
              imgRepeated,
              getConfig({
                resourcesUrl: this.options.resourcesUrl,
                debug: this.debug,
                iris: false,
                live: false
              })
            );

            if (fianlResult.face && fianlResult.face.length > 0) {
              // emit data
              subscribe.next({ photo, rectPhoto, result });

              // if send , unsubscibe it 
              if (photo) {
                sub.unsubscribe();
                subscribe.complete();
              }
            }
```
my code,photo is the base64 of current photo.

**Comment by Howjune** [comment]
> 
> 
> `result.face.confidence` is just a pointer to `boxConfidence` if mesh is disabled or to `faceConfidence` if mesh is enabled.

Yes. but none face is still get high score.

**Comment by vladmandic** [comment]
> Yes. but none face is still get high score.

Can you post values for `boxConfidence` AND `faceConfidence`? 

> Human: could not convert input to tensor

That's a separate issue and it's due to fact that image is not ready at the time when you call `human`.
Basically, you assign `image.src` and then call `human.detect` immediately after, but browser may take some time to actually process the src you just assigned  

What you need is to wait for image to actually get loaded before calling `human` - something like:
````js
img.onload = async () => {
   const result = await human.detect(image);
};
img.src = 'some_string...';

**Comment by Howjune** [comment]
Oh.I see. That's my fault.I didn't consider that.lol. Thank you so much.
The value of faceConfidence maybe around 0.7.

**Comment by Howjune** [comment]
I hope to recognize faces more accurately even if the mesh is disabled.The reason why I disable it is that it spends more time to load. The user experience is not particularly good

**Comment by Howjune** [comment]
And the browser will caton before runing the scripts.

**Comment by Howjune** [comment]
> And does it drop after a while

Sorry i didn't notice it.

**Comment by vladmandic** [comment]
if you only need face detection and nothing else, you can try alternative model  
default model is BlazeFace, but that only works when combined with FaceMesh  
but there is also FaceBoxes which is a standalone detector - exactly for the reason that BlazeFace is not that good when running as standalone model  

```js
const myConfig = {
  videoOptimized: false,
  face: {
    enabled: true,
    detector: { modelPath: '/human/models/faceboxes.json', enabled: true, minConfidence: 0.5 },
    mesh: { enabled: false },
    iris: { enabled: false },
    age: { enabled: false },
    gender: { enabled: false },
    emotion: { enabled: false },
  },
  body: { enabled: false },
  hand: { enabled: false },
};
```js

I'm still looking for a single good algorithm for face detection that can replace both.

**Comment by Howjune** [comment]
> I just noticed that you have `videoOptimized: true` in your config meaning actual face detection is skipped for some frames and instead it returns interpolated values (for performance reasons)
> 
> When testing, it's best to force full detection each frame by setting `videoOptimized: false`

I will do as u say.

**Comment by vladmandic** [comment]
> I hope to recognize faces more accurately even if the mesh is disabled.The reason why I disable it is that it spends more time to load. The user experience is not particularly good

Regarding load time - is it actual **load** or **initialization** that is not good for user experience?
You can completely skip initialization time if you set `backend: 'wasm'` (instead of default `webgl` as `webgl` takes time to initialize GL shaders, but WASM has no such thing). Once initialized, `webgl` is faster than `wasm`, but for just face detection that wouldn't matter.

**Comment by vladmandic** [comment]
To summarize, 
- Try using `backend: wasm` to eliminate initialization time
- Wait for image onload before processing it
- Set `videoOptimized: false` to avoid false cached results and then report results for `boxConfidence` and `faceConfidence`
- Try using alternative model (faceboxes instead of blazeface)

**Comment by Howjune** [comment]
> To summarize,
> 
> * Try using `backend: wasm` to eliminate initialization time
> * Wait for image onload before processing it
> * Set `videoOptimized: false` to avoid false cached results and then report results for `boxConfidence` and `faceConfidence`
> * Try using alternative model (faceboxes instead of blazeface)
I will do it. It's so cool to discuss with you. I like the repo which is wonderful.

**Comment by vladmandic** [comment]
I just published an update:
- boxConfidence was not correctly updated when mesh is disabled, so filtering for low scores did not work as expected
  I've tried with new code and minConfidence set to 0.05 (so plenty of false results) and it works fine now with  
  face covered with hand, etc. - scores get reduced accordingly
- Code was doing unnecessary calculations when mesh is disabled  
  (e.g., expensive rotation matrix when there is no mesh that actually uses them
  Fixed this as well, so code with mesh disabled should be a lot faster

**Comment by Howjune** [comment]
Thanks a lot.I will try it now.

**Comment by vladmandic** [comment]
You can skip testing faceboxes, I've decided to remove it and focus on improving primary model - blazeface. Everything else stands.

**Comment by Howjune** [comment]
![image](https://user-images.githubusercontent.com/43192276/110568219-a3002680-818d-11eb-8324-5ab2dfd2d4f2.png)
My fist was recognized as a face.

``` typescript
backend: 'wasm', 
human/models/blazeface-back.json
```

**Comment by vladmandic** [comment]
With mesh disabled?
What's the boxConfidence?


How about with mesh enabled? And what's the faceConfidence then?

**Comment by Howjune** [comment]
It work well when mesh is enable and I disabled the mesh and the value of confidence is 0.5xxx.

**Comment by vladmandic** [comment]
Glad it works with mesh.

And even without mesh, the score drops, so setting minimum score to something like 0.75 would solve both cases.

Anyhow, I'll take another look to tweak scoring of model without mesh as it does seem a bit high.

**Comment by Howjune** [comment]
It is faster than before when i used WASM and with mesh enabled.I guess u have solved my problem already.
Thanks！(:

**Comment by Howjune** [comment]
It is faster than before when i used WASM and with mesh enabled.I guess u have solved my problem already.
Thanks！(:(:

**Comment by MuriloRS** [comment]
> if you only need face detection and nothing else, you can try alternative model
> default model is BlazeFace, but that only works when combined with FaceMesh
> but there is also FaceBoxes which is a standalone detector - exactly for the reason that BlazeFace is not that good when running as standalone model
> 
> ```js
> const myConfig = {
>   videoOptimized: false,
>   face: {
>     enabled: true,
>     detector: { modelPath: '/human/models/faceboxes.json', enabled: true, minConfidence: 0.5 },
>     mesh: { enabled: false },
>     iris: { enabled: false },
>     age: { enabled: false },
>     gender: { enabled: false },
>     emotion: { enabled: false },
>   },
>   body: { enabled: false },
>   hand: { enabled: false },
> };
> ```
> 
> I'm still looking for a single good algorithm for face detection that can replace both.

Where can I find this faceboxes model?

**Comment by vladmandic** [comment]
I wrote a month ago:
> You can skip testing faceboxes, I've decided to remove it and focus on improving primary model - blazeface

If you really want to check it out, you can clone the repository and go back to version 0.9.x (`faceboxes` model was removed with 1.0 release). alternative, old versions are archived on `npmjs`, so you can just install a 0.9.x version of `human` using `npm`.
